src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs

```diff
@@ -0,0 +1,361 @@
+
+                    _textBox.PropertyChanged += _textBox_PropertyChanged;
+                    _textBox.LayoutUpdated += _textBox_LayoutUpdated;
+                }
+            }
+            else
+            {
+                if (_textBox != null)
+                {
+                    _textBox.RemoveHandler(TextBox.TextChangingEvent, TextChanged);
+                    _textBox.RemoveHandler(KeyDownEvent, TextBoxKeyDown);
+                    _textBox.RemoveHandler(PointerReleasedEvent, TextBoxPointerReleased);
+                    _textBox.RemoveHandler(LostFocusEvent, TextBoxLostFocus);
+                    _textBox.RemoveHandler(Gestures.HoldingEvent, TextBoxHolding);
+
+                    _textBox.PropertyChanged -= _textBox_PropertyChanged;
+                    _textBox.LayoutUpdated -= _textBox_LayoutUpdated;
+                }
+
+                _textBox = null;
+            }
+        }
+
+        private void TextBoxHolding(object? sender, HoldingRoutedEventArgs e)
+        {
+            if (ShowContextMenu())
+                e.Handled = true;
+        }
+
+        internal bool ShowContextMenu()
+        {
+            if (_textBox != null)
+            {
+                if (_textBox.ContextFlyout is MenuFlyout flyout)
+                {
+                    var verticalOffset = (double.IsNaN(_textBox.LineHeight) ? _textBox.FontSize : _textBox.LineHeight) + ContextMenuPadding;
+
+                    TextSelectionHandle? handle = null;
+
+                    if (_textBox.SelectionStart != _textBox.SelectionEnd)
+                    {
+                        if (_startHandle.IsEffectivelyVisible)
+                            handle = _startHandle;
+                        else if (_endHandle.IsEffectivelyVisible)
+                            handle = _endHandle;
+                    }
+                    else
+                    {
+                        if (_caretHandle.IsEffectivelyVisible)
+                        {
+                            handle = _caretHandle;
+                        }
+                    }
+
+                    if (handle != null)
+                    {
+                        var topLeft = ToTextBox(handle.GetTopLeft());
+                        flyout.VerticalOffset = topLeft.Y - verticalOffset;
+                        flyout.HorizontalOffset = topLeft.X;
+                        flyout.Placement = PlacementMode.Top;
+                        flyout.ShowAt(_textBox);
+
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private void TextBoxPointerReleased(object? sender, PointerReleasedEventArgs e)
+        {
+            if (e.Pointer.Type != PointerType.Mouse)
+            {
+                ShowHandles = true;
+
+                MoveHandlesToSelection();
+                EnsureVisible();
+            }
+        }
+
+        private void _textBox_LayoutUpdated(object? sender, EventArgs e)
```

Comment:
Nit: naming `TextBoxLayoutUpdated` and `TextBoxPropertyChanged` to match the other handlers.